sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,251 @@
+	SubjectUid uint32
+
+	// Path is the path of the file, as seen by the process triggering the request.
+	Path string
+	// Permission is the opaque permission that is being requested.
+	Permission interface{}
+	// YesNo is a channel for writing the response.
+	YesNo chan bool
+}
+
+func newRequest(l *Listener, msg *notify.MsgNotificationFile) *Request {
+	var perm interface{}
+	if msg.Class == notify.AA_CLASS_FILE {
+		_, deny, _ := msg.DecodeFilePermissions()
+		perm = deny
+	}
+	return &Request{
+		l: l, // why is this needed?
+
+		Pid:        msg.Pid,
+		Label:      msg.Label,
+		Path:       msg.Name,
+		SubjectUid: msg.SUID,
+
+		Permission: perm,
+
+		YesNo: make(chan bool, 1),
+	}
+}
+
+var (
+	// ErrListenerNotSupported indicates that the kernel does not support apparmor prompting
+	ErrListenerNotSupported = errors.New("kernel does not support apparmor notifications")
+
+	osOpen      = os.Open
+	notifyIoctl = notify.Ioctl
+)
+
+// Register opens and configures the apparmor notification interface.
+//
+// If the kernel does not support the notification mechanism the error is ErrNotSupported.
+func Register() (*Listener, error) {
+	path := notify.SysPath
+	if override := os.Getenv("PROMPT_NOTIFY_PATH"); override != "" {
+		path = override
+	}
+
+	notifyFile, err := osOpen(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrListenerNotSupported
+		}
+		return nil, fmt.Errorf("cannot open %q: %v", path, err)
+	}
+
+	msg := notify.MsgNotificationFilter{ModeSet: notify.APPARMOR_MODESET_USER}
+	data, err := msg.MarshalBinary()
+	if err != nil {
+		notifyFile.Close()
+		return nil, err
+	}
+	ioctlBuf := notify.BytesToIoctlRequestBuffer(data)
+	_, err = notifyIoctl(notifyFile.Fd(), notify.APPARMOR_NOTIF_SET_FILTER, ioctlBuf)
+	if err != nil {
+		notifyFile.Close()
+		return nil, fmt.Errorf("cannot notify ioctl to modeset user on %q: %v", path, err)
+	}
+
+	poll, err := epoll.Open()
+	if err != nil {
+		notifyFile.Close()
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %v", err)
+	}
+	// XXX: Do we need a notification for Writable, to send responses back?
+	if err := poll.Register(int(notifyFile.Fd()), epoll.Readable); err != nil {
+		notifyFile.Close()
+		poll.Close()
+		return nil, fmt.Errorf("cannot register epoll on %q: %v", path, err)
+	}
+
+	listener := &Listener{
+		R: make(chan *Request),
+		E: make(chan error),
+
+		notifyFile: notifyFile,
+		poll:       poll,
+	}
+	return listener, nil
+}
+
+func (l *Listener) decodeAndDispatchRequest(buf []byte, tomb *tomb.Tomb) error {
+	var nmsg notify.MsgNotification
+	if err := nmsg.UnmarshalBinary(buf); err != nil {
+		return err
+	}
+	// What kind of notification message did we get?
+	switch nmsg.NotificationType {
+	case notify.APPARMOR_NOTIF_OP:
+		var omsg notify.MsgNotificationOp
+		if err := omsg.UnmarshalBinary(buf); err != nil {
+			return err
+		}
+		// What kind of operation notification did we get?
+		switch omsg.Class {
+		case notify.AA_CLASS_FILE:
```

Comment:
I really like the idea separate handler functions for each type -- especially coupled with the other de-nesting above, this makes `decodeAndDispatchRequest` much cleaner. I want to keep the switch statement here, since we are planning to support other types of notification classes in the not-so-distant future.